common/tasks/multi_destination_scheduler.go

```diff
@@ -0,0 +1,241 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tasks
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+)
+
+type RateLimiter interface {
+	Wait(context.Context) error
+}
+
+// LimiterOptions are options for a single destination's limiter.
+type LimiterOptions struct {
+	// Concurrency returns the concurrency limit.
+	Concurrency func() int
+	RateLimiter RateLimiter
+}
+
+// limiter acts a concurrency and rate limiter for a single destination.
+// All access to the limiter must be locked.
+type limiter[T Task] struct {
+	sync.Mutex
+	options LimiterOptions
+	// Number of reservations help by this limiter. Restricts concurrency.
+	reservations int
+	// Any tasks that exceed the concurrency limitation are buffered in this slice.
+	buffer []T
+}
+
+func newLimiter[T Task](options LimiterOptions) *limiter[T] {
+	return &limiter[T]{
+		options: options,
+	}
+}
+
+// enqueue a task in the limiter's buffer.
+func (lim *limiter[T]) enqueue(task T) {
+	lim.buffer = append(lim.buffer, task)
+}
+
+// dequeue a task from the limiter's buffer.
+func (lim *limiter[T]) dequeue() (task T, ok bool) {
+	if len(lim.buffer) > 0 {
+		task, lim.buffer = lim.buffer[0], lim.buffer[1:]
+		return task, true
+	}
+	return task, false
+}
+
+// reserve a concurrency slot. Returns true on success.
+func (lim *limiter[T]) reserve() bool {
+	if lim.reservations < lim.options.Concurrency() {
+		lim.reservations++
+		return true
+	}
+	return false
+}
+
+// release a reservation.
+func (lim *limiter[T]) release() {
+	lim.reservations--
+}
+
+// MultiDestinationSchedulerOptions are options for creating a [MultiDestinationScheduler].
+type MultiDestinationSchedulerOptions[K comparable, T Task] struct {
+	// A function to determine the destination of a task.
+	KeyFn func(T) K
+	// Limiter options for a given destination.
+	LimiterOptionsFn func(K) LimiterOptions
```

Comment:
I think that what we actually want is a scheduler factory here. If the `MultiDestinationScheduler` is reused for a different key type, it's possible it also wants to have different scheduling logic for each destination. In addition, that makes this type follow the single-responsibility principle: it's only responsible for sending tasks to another scheduler depending on that task's "partition".